webui/src/pages/auth/users/index.jsx

```diff
@@ -1,51 +1,129 @@
     FormattedDate,
     Loading,
     RefreshButton
 } from "../../../lib/components/controls";
 import validator from "validator/es";
-import { disallowPercentSign, INVALID_USER_NAME_ERROR_MESSAGE } from "../validation";
-import { resolveUserDisplayName } from "../../../lib/utils";
+import {disallowPercentSign, INVALID_USER_NAME_ERROR_MESSAGE} from "../validation";
+import {resolveUserDisplayName} from "../../../lib/utils";
+import {SearchIcon} from "@primer/octicons-react";
+import InputGroup from "react-bootstrap/InputGroup";
+import {allUsersFromLakeFS} from "../../../lib/components/auth/users";
+import {useRouter} from "../../../lib/hooks/router";
 
+const DEFAULT_DEBOUNCE_MS = 300;
+const DEFAULT_LISTING_AMOUNT = 100;
+const DECIMAL_RADIX = 10;
 const USER_NOT_FOUND = "unknown";
 export const GetUserEmailByIdContext = createContext();
 
 
-const UsersContainer = ({nextPage, refresh, setRefresh, error, loading, userListResults}) => {
+const UsersContainer = ({ refresh, setRefresh, allUsers, loading, error }) => {
     const { user } = useUser();
     const currentUser = user;
 
     const router = useRouter();
-    const after = (router.query.after) ? router.query.after : "";
     const [selected, setSelected] = useState([]);
     const [deleteError, setDeleteError] = useState(null);
     const [showCreate, setShowCreate] = useState(false);
     const [showInvite, setShowInvite] = useState(false);
-    
-    
 
-    useEffect(() => { setSelected([]); }, [refresh, after]);
+    const routerSearchPrefix = (router.query.searchPrefix) ? router.query.searchPrefix : "";
+    const routerUserIdxFromAllUsers = (router.query.userIdxFromAllUsers)
+        ? parseInt(router.query.userIdxFromAllUsers, DECIMAL_RADIX) : 0;
+    const routerHasMorePages = (router.query.hasMorePages) ? router.query.hasMorePages : "false";
+
+    const [searchInput, setSearchInput] = useState(routerSearchPrefix);
+    const [searchPrefix, setSearchPrefix] = useState(routerSearchPrefix);
+    const [prevSearchPrefix, setPrevSearchPrefix] = useState(searchPrefix);
+    const [prevAllUsersLen, setPrevAllUsersLen] = useState(0);
+    const [userIdxFromAllUsers, setUserIdxFromAllUsers] = useState(routerUserIdxFromAllUsers);
+    const [hasMorePages, setHasMorePages] = useState(routerHasMorePages);
+    const [filteredUsers, setFilteredUsers] = useState([]);
+
+
+    useEffect(() => { setSelected([]); }, [refresh]);
 
     const authCapabilities = useAPI(() => auth.getAuthCapabilities());
+
+    useEffect(() => {
+        if (searchInput !== searchPrefix) {
+            // Update the search prefix if the user hasn't typed for 300ms.
+            const handler = setTimeout(() => {
+                setPrevSearchPrefix(searchPrefix);
+                setSearchPrefix(searchInput);
+                setUserIdxFromAllUsers(0);
+            }, DEFAULT_DEBOUNCE_MS);
+            return () => clearTimeout(handler);
+        }
+    }, [searchInput]);
+
+    useEffect(() => {
```

Comment:
This part seems too complex to understand without prior knowledge.
In such cases, I suggest:
1. Extract this logic into sub-functions. This will force you to name these parts of the flow - it'll be more descriptive for others that read this, and will make you consider which parts are really necessary.
2. Try to think, where can you simplify? Are all the variables really needed? All `if`s?
3. As a last resort, add comments. But it's probably needed only in some very delicate parts, details or algorithms. I don't think that's the case here - but maybe some details about the logic behind fetching all the users, and how you use the idx for paginating the results.